app/models/patient.rb

```diff
@@ -53,6 +53,11 @@ class Patient < ApplicationRecord
 
   attribute :call_result, :string
 
+  scope :follow_ups, -> (period, last: nil) {
```

Comment:
This scope is intentionally partially applied in two ways:

1. It doesn't join over `blood_pressures`, even though it expects it to exist
2. It ends with a group clause, which means the caller must either add more groups or run an aggregate themselves.

This is done for 2 reasons:

1. Rails doesn't properly resolve queries when, say, this scope is run through [facilities](https://github.com/simpledotorg/simple-server/pull/888/files#diff-a0e971b921f56b979cc5aa07fee7066fR52) like:

`Facility.first.patient_follow_ups.count`

Now, if we had a an extra `joins(:blood_pressures)` in `Patient.follow_ups`. It would end up sticking two `INNER JOINs` with `BloodPressures` because patients in Facility are already [through](https://github.com/simpledotorg/simple-server/pull/888/files#diff-a0e971b921f56b979cc5aa07fee7066fR20) `blood_pressures`.

Double joins still result in correct data, but make the query performance quadratic. The usual difference (on SBX) that we were seeing was like: 1.5s vs 12s.

The unsavoury result of this is if you're using this method from Patient class, you'd have to do `Patient.joins(:blood_pressures).follow_ups`

2. The group-by clause is not technically meaningless. This scope still allows you to chain more things in front. It just won't provide you with a realized result immediately. It'd expect some aggregate fn, more group-bys, or wheres and other SQLs.

Reply:
I did a lot of digging to see if there's a way to say "please join this if not already joined", but looks like that concept is frowned upon because it messes up some internal table alias assignment process.

I think this is fine, good balance of simplicity and performance.